src/lib.rs

```diff
@@ -176,457 +138,684 @@ impl FromStr for ZKConnectString {
+                            Loop::Continue(WatchLoopState {
+                                watcher,
+                                curr_event: event,
+                                delay:WATCH_LOOP_NODELAY
+                            })
+                        },
+                        //
+                        // If we didn't get a valid event, this means the Stream
+                        // got closed, which indicates a connection issue, so we
+                        // reconnect.
+                        //
+                        None => {
+                            error!(log, "Event stream closed; reconnecting");
+                            Loop::Break(NextAction::Reconnect(
+                                RECONNECT_NODELAY))
+                        }
+                    };
+                    ok(loop_next)
+                })
+                .or_else(move |_| {
+                    //
+                    // If we get an error from the event Stream, we assume that
+                    // something went wrong with the zookeeper connection and
+                    // attempt to reconnect.
+                    //
+                    // The stream's error type is (), so there's no information
+                    // to extract from it.
+                    //
+                    error!(oe_log, "Error received from event stream");
+                    ok(Loop::Break(NextAction::Reconnect(RECONNECT_NODELAY)))
+                }))
+        })
+        //
+        // If some error occurred getting the data, we assume we should
+        // reconnect to the zookeeper server.
+        //
+        .or_else(move |error| {
+            error!(oe_log, "Error getting data"; "error" => LogItem(error));
+            ok(Loop::Break(NextAction::Reconnect(RECONNECT_NODELAY)))
+        })
+    })
+    .map_err(|e| panic!("delay errored; err: {:?}", e))
 }
 
-impl Watcher for SessionEventWatcher {
-    fn handle(&self, e: WatchedEvent) {
-        println!("{:?}", e);
-        // Zookeeper session state changes should have WatchedEventType::None,
-        // so we assert this fact
-        match e.event_type {
-            WatchedEventType::None =>
-                // This is an internal channel that we manage, so we expect
-                // sending to succeed
-                self.event_tx.send(ResolverEvent::SessionEvent).unwrap(),
-            _ => panic!("Unexpected event type: {:?}", e.event_type)
-        };
-    }
+fn connect_loop(
+    pool_tx: Sender<BackendMsg>,
+    last_backend: Arc<Mutex<Option<BackendKey>>>,
+    connect_string: ZkConnectString,
+    cluster_state_path: String,
+    delay: Duration,
+    log: Logger
+) -> impl Future<Item = Loop<(), Duration>,
+    Error = ()> + Send {
+
+    let oe_log = log.clone();
+    Delay::new(Instant::now() + delay)
+    .and_then(move |_| {
+        info!(log, "Connecting to ZooKeeper cluster");
+        ZooKeeper::connect(&connect_string.to_string().parse().unwrap())
```

Comment:
Done. I added an empty-string check to `FromStr` and consequently had to introduce a new error type specific to `ZkConnectString`, because apparently `AddrParseError`s can't be constructed outside of the module in which they're defined.

I also chose to make the new `get_addr_at()` function for `ZkConnectString` return None upon out-of-bounds access rather than panic, so the caller will still have to unwrap it in the line commented on above.

Reply:
I think that's a good choice here, but just to be extra cautious if something does go wrong can we use `expect` with a contextual message instead of `unwrap` for the `get_addr_at` result?